bundles/org.openhab.binding.mideaac/src/main/java/org/openhab/binding/mideaac/internal/discovery/MideaACDiscoveryService.java

```diff
@@ -0,0 +1,370 @@
+        } catch (IOException e) {
+            logger.debug("Error during discovery: {}", e.getMessage());
+        } finally {
+            closeDiscoverSocket();
+        }
+    }
+
+    /**
+     * Opens a {@link DatagramSocket} and sends a packet for discovery of Midea AC devices.
+     *
+     * @throws SocketException
+     * @throws IOException
+     */
+    private void startDiscoverSocket() throws SocketException, IOException {
+        startDiscoverSocket("255.255.255.255", null);
+    }
+
+    /**
+     * Start the discovery Socket
+     * 
+     * @param ipAddress IP Address
+     * @param discoveryHandler Discovery handler
+     * @throws SocketException Socket Exception
+     * @throws IOException IO Exception
+     */
+    public void startDiscoverSocket(String ipAddress, @Nullable DiscoveryHandler discoveryHandler)
+            throws SocketException, IOException {
+        logger.trace("Discovering: {}", ipAddress);
+        this.discoveryHandler = discoveryHandler;
+        discoverSocket = new DatagramSocket(new InetSocketAddress(Connection.MIDEAAC_RECEIVE_PORT));
+        DatagramSocket discoverSocket = this.discoverSocket;
+        discoverSocket.setBroadcast(true);
+        discoverSocket.setSoTimeout(udpPacketTimeout);
+        final InetAddress broadcast = InetAddress.getByName(ipAddress);
+        {
+            final DatagramPacket discoverPacket = new DatagramPacket(CommandBase.discover(),
+                    CommandBase.discover().length, broadcast, Connection.MIDEAAC_SEND_PORT1);
+            discoverSocket.send(discoverPacket);
+            if (logger.isTraceEnabled()) {
+                logger.trace("Broadcast discovery package sent to port: {}", Connection.MIDEAAC_SEND_PORT1);
+            }
+        }
+        {
+            final DatagramPacket discoverPacket = new DatagramPacket(CommandBase.discover(),
+                    CommandBase.discover().length, broadcast, Connection.MIDEAAC_SEND_PORT2);
+            discoverSocket.send(discoverPacket);
+            if (logger.isTraceEnabled()) {
+                logger.trace("Broadcast discovery package sent to port: {}", Connection.MIDEAAC_SEND_PORT2);
+            }
+        }
+    }
+
+    /**
+     * Closes the discovery socket and cleans the value. No need for synchronization as this method is called from a
+     * synchronized context.
+     */
+    private void closeDiscoverSocket() {
+        DatagramSocket discoverSocket = this.discoverSocket;
+        if (discoverSocket != null) {
+            discoverSocket.close();
+            discoverSocket = null;
+        }
+    }
+
+    /**
+     * Register a device (thing) with the discovered properties.
+     *
+     * @param packet containing data of detected device
+     */
+    private void thingDiscovered(DatagramPacket packet) {
+        DiscoveryResult dr = discoveryPacketReceived(packet);
+        DiscoveryHandler discoveryHandler = this.discoveryHandler;
+        if (dr != null) {
```

Comment:
(null check is perfect) Ultra minor improvmeent, just assign it after the condition, to prevend assignment in the rare case it is not needed.
```suggestion
        if (dr != null) {
            DiscoveryHandler discoveryHandler = this.discoveryHandler;
```

Reply:
Done